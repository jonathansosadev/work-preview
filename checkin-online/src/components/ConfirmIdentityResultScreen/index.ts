export {ConfirmIdentityResultScreen as default} from './ConfirmIdentityResultScreen';
