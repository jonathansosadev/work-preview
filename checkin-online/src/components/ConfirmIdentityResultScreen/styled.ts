import styled from 'styled-components';
import {default as BaseProgressCircle} from '../ProgressCircle';
import {device} from '../../styled/device';
import {OriginalButton} from '../Button';

export const Content = styled.div`
  text-align: center;
  cursor: default;
`;

export const Image = styled.img`
  width: 84px;
  height: 84px;

  @media (max-width: ${device.tablet}) {
    margin-top: 27px;
  }
`;

export const Title = styled.div`
  font-family: ProximaNova-Bold, sans-serif;
  color: #161643;
  margin-top: 17px;
  font-size: 25px;
`;

export const Subtitle = styled.div`
  font-family: ProximaNova-Light, sans-serif;
  font-size: 17px;
  color: #161643;
  max-width: 317px;
  margin: 0 auto;
`;

export const ButtonWrapper = styled.div`
  margin: 80px 0 35px;
  display: flex;
  justify-content: center;

  @media (max-width: ${device.tablet}) {
    margin-top: 70px;
  }
`;

export const ScanResultContainer = styled.div`
  width: 317px;
  position: relative;
  margin: 40px auto 0;
  min-height: 180px;
  user-select: none;
`;

export const DocumentResultArea = styled.div`
  position: absolute;
  left: 5px;
  top: 24px;
  display: flex;
  width: 113px;
  flex-direction: column;
  font-family: ProximaNova-Semibold, sans-serif;
  font-size: 17px;
  color: #161643;
  align-items: center;
`;

type ResultImageProps = {
  image: string;
  mirror: boolean;
};
export const ResultImage = styled.div<ResultImageProps>`
  background-image: url(${props => props.image});
  transform: ${props => props.mirror && 'rotateY(180deg)'};
  box-shadow: 0 16px 16px #2699fb1a;
  width: 100%;
  height: 113px;
  border: 5px solid white;
  background-color: white;
  border-radius: 50%;
  margin-bottom: 13px;
  background-size: cover;
  background-position: center;
`;

export const SelfieResultArea = styled(DocumentResultArea)`
  left: unset;
  right: 5px;
`;

export const ProgressArea = styled.div`
  position: absolute;
  left: 0;
  right: 0;
  top: 0;
  bottom: 0;
  margin: auto;
  z-index: 10;
`;

export const ProgressCircle = styled(BaseProgressCircle)`
  margin: auto;
`;

export const TitleIcon = styled.img`
  height: 25px;
  max-width: 25px;
  margin-left: 7px;
  vertical-align: middle;
`;

export const TitleSuccessIcon = styled(TitleIcon)`
  margin-top: -6px;
`;

export const TitleErrorIcon = styled(TitleIcon)``;

export const TryButton = styled(OriginalButton)`
  margin: 0 auto 50px;
`;
