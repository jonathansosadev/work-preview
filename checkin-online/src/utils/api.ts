import * as Sentry from '@sentry/react';
import {getAuthAndBaseHeaders} from '../api';

const FAILURE_RESPONSE_STATUSES = [400, 401, 429];

function stringifySentryRequestInit(options: any) {
  const base64Fields = ['signature', 'picture_file', 'front_side_scan', 'back_side_scan'];
  const body = options?.body;

  if (body) {
    const parsedBody = JSON.parse(body);

    base64Fields.forEach(field => {
      if (parsedBody[field]) {
        parsedBody[field] = 'base64';
      }
    });

    return JSON.stringify({
      ...options,
      body: parsedBody,
    });
  }

  return JSON.stringify(options);
}

function getResponseHeaders(response?: Response) {
  let headers = '';

  if (!response) {
    return headers;
  }

  // @ts-ignore
  for (const pair of response?.headers?.entries()) {
    headers += `${pair[0]}: ${pair[1]}; `;
  }
  return headers;
}

function getAndHandleResponseError(response?: Response, body?: any, options?: any) {
  if (!response) {
    return null;
  }

  if (FAILURE_RESPONSE_STATUSES.includes(response.status)) {
    return body;
  }

  if (response.status === 401) {
    Sentry.captureMessage(
      `Error 401: ${response.url}. Headers: ${JSON.stringify(
        getAuthAndBaseHeaders(),
      )}. Request init: ${stringifySentryRequestInit(options)}`,
    );
  }

  if (response.status === 404) {
    Sentry.captureMessage(
      `Error 404: ${response.url}. Request init: ${stringifySentryRequestInit(options)}`,
    );

    return {
      message: `404 Not found. Please contact support.`,
      body,
    };
  }

  if (response.status === 405) {
    Sentry.captureMessage(
      `Error 405: ${response.url}. Request init: ${stringifySentryRequestInit(options)}`,
    );

    return {
      message: `405 Method not allowed. Please contact support.`,
      body,
    };
  }

  if (response.status === 500) {
    Sentry.captureMessage(
      `Error 500: ${response.url}. Request init: ${stringifySentryRequestInit(options)}`,
    );

    return {
      message: '500 Internal Server error. Please contact support.',
      body,
    };
  }

  if (response.status === 502) {
    Sentry.captureMessage(
      `Error 502: ${response.url}. Request init: ${stringifySentryRequestInit(options)}`,
    );

    return {
      message: '502 Bad Gateway. Please contact support.',
      body,
    };
  }

  if (response.status === 503) {
    Sentry.captureMessage(
      `Error 503: ${response.url}. Request init: ${stringifySentryRequestInit(options)}`,
    );

    return {
      message: '503 Service Unavailable. Please contact support.',
      body,
    };
  }

  if (response.status === 504) {
    Sentry.captureMessage(
      `Error 504: ${response.url}. Request init: ${stringifySentryRequestInit(options)}`,
    );

    return {
      message: '504 Gateway Timeout. Please contact support.',
      body,
    };
  }

  return null;
}

export {getAndHandleResponseError, getResponseHeaders};
