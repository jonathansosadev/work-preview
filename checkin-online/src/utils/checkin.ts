import {ComputedReservationDetailsContextProps} from '../context/computedReservationDetails';
import {PATHS} from '../Routes';

type FinishURLProps = {
  reservationDetails: ComputedReservationDetailsContextProps;
  locationState?: any;
  skip?: {
    hasUpselling: boolean;
  };
};

function getFinishCheckinHistoryArgs({
  reservationDetails,
  locationState,
  skip,
}: FinishURLProps) {
  if (!skip?.hasUpselling && reservationDetails.hasUpselling) {
    return {
      url: '/deals-list',
      state: locationState,
    };
  }

  if (reservationDetails.hasTaxes) {
    return {
      url: PATHS.taxesSetup,
      state: locationState,
    };
  }

  if (reservationDetails.isSomePayments) {
    return {
      url: PATHS.payments,
      state: locationState,
    };
  }

  if (reservationDetails.hasDeposits) {
    return {
      url: '/deposits/payment',
      state: locationState,
    };
  }

  return {
    url: '/finish',
    state: locationState,
  };
}

export {getFinishCheckinHistoryArgs};
