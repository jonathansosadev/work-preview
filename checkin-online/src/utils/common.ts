import i18next from 'i18next';
import moment from 'moment';
import {COUNTRY_CODES} from './constants';

export type ErrorType =
  | {
      message: string;
      errors: any;
      details: string;
    }
  | string;

function getErrorMessage(error: ErrorType) {
  let message = '';

  if (!error) {
    return message;
  }

  if (typeof error === 'string') {
    return error;
  }

  if (error?.details) {
    return error?.details;
  }
  if (error?.errors?.length) {
    const nonFieldErrors = 'non_field_errors';
    Object.values(error.errors).forEach((err: any) => {
      message += ` ${err.field && err.field !== nonFieldErrors ? `${err.field}:` : ''} ${
        err.message
      }\n`;
    });
  } else if (error?.message) {
    message = error.message;
  } else {
    if (typeof error === 'object') {
      Object.values(error).forEach(err => {
        message += ` ${err}\n`;
      });
    }
  }

  if (typeof message === 'object') {
    return '';
  }

  if (message?.startsWith('Unexpected JSON')) {
    return 'Server error.';
  }

  return message;
}

function getURLSearchParam(name: string, source = window.location.search) {
  const searchParams = new URLSearchParams(source);
  return searchParams.get(name);
}

function setURLSearchParam(name: string, value = '', target = window.location.search) {
  const searchParams = new URLSearchParams(target);
  return searchParams.set(name, value);
}

function deleteQueryParam(name: string, source = window.location.search) {
  const searchParams = new URLSearchParams(source);
  searchParams.delete(name);

  return searchParams.toString();
}

function parseJwt(token: string) {
  const base64Url = token.split('.')[1];
  const base64 = base64Url.replace(/-/g, '+').replace(/_/g, '/');
  const jsonPayload = decodeURIComponent(
    atob(base64)
      .split('')
      .map(function (c) {
        return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
      })
      .join(''),
  );

  return JSON.parse(jsonPayload);
}

function getCurrentLocale() {
  try {
    return i18next.language.slice(0, 2);
  } catch (e) {
    return 'en';
  }
}

function getEnvVariable(key: string) {
  const variable = process.env[key];

  if (!variable) {
    throw new Error(`${key} .env variable is missing.`);
  }

  return variable;
}

function getMinCheckInDate(countryCode?: string) {
  let date;

  if (countryCode === COUNTRY_CODES.portugal) {
    date = moment().subtract(4, 'days');
  } else {
    date = moment().subtract(1, 'days').startOf('day');
  }

  return date;
}

function copyToClipboard(value: string) {
  const el = document.createElement('textarea');
  el.value = value;
  el.setAttribute('readonly', '');
  el.style.position = 'absolute';
  el.style.left = '-9999px';
  document.body.appendChild(el);
  el.select();
  document.execCommand('copy');
  document.body.removeChild(el);
}

function getBase64(file: File): Promise<any> {
  return new Promise((resolve, reject) => {
    if (!file) {
      return reject('File is missing');
    }
    const reader = new FileReader();
    reader.readAsDataURL(file);
    reader.onload = () => resolve(reader.result);
    reader.onerror = error => reject(error);
  });
}

export {
  getBase64,
  getErrorMessage,
  getURLSearchParam,
  setURLSearchParam,
  getCurrentLocale,
  parseJwt,
  getEnvVariable,
  getMinCheckInDate,
  deleteQueryParam,
  copyToClipboard,
};
