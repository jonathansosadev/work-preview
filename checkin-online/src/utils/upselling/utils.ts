import {Offer} from './types';

function getSummaryPrices(offer: Offer) {
  return offer.items.reduce((sum, itemPrice) => sum + Number(itemPrice?.price || 0), 0);
}

export {getSummaryPrices};
