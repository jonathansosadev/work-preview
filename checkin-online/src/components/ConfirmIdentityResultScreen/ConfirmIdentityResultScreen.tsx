import React from 'react';
import {useTranslation} from 'react-i18next';
import {useLocation, useHistory} from 'react-router-dom';
import {useBiomatchUpdateGuest} from '../../hooks/useBiomatchUpdateGuest';
import {useComputedReservationDetails} from '../../context/computedReservationDetails';
import {breakIntoLines} from '../../utils/translations';
import {isMobileOrTablet} from '../../utils/mobile';
import {getIdentityVerificationHeaderProps} from '../ConfirmIdentityGuideScreen';
import scanningIcon from '../../assets/scanning_icn.svg';
import successSmallIcon from '../../assets/success-small.svg';
import errorSmallIcon from '../../assets/error-small.svg';
import Header from '../Header';
import Button from '../Button';
import {
  ScanResultContainer,
  Title,
  Subtitle,
  Image,
  Content,
  ButtonWrapper,
  DocumentResultArea,
  SelfieResultArea,
  ProgressCircle,
  ProgressArea,
  TitleErrorIcon,
  TitleSuccessIcon,
  ResultImage,
  TryButton,
} from './styled';

const matchingScaleCoefficient = 0.35;

function getFaceMatchPercentage(distance = 0.0) {
  const percentage = 100 - (1 - Number(distance) / (1 - matchingScaleCoefficient)) * 100;

  if (percentage < 0) {
    return 0;
  }

  if (percentage > 100) {
    return 100;
  }

  return percentage;
}

type LocationTypes = {
  facesComparingResult: any;
  documentPhoto: string;
  selfiePhoto: string;
  documentCheckId: string;
  selfieCheckId: string;
  failedOCRAttempts?: number;
  biomatch?: {
    biomatch_passed: boolean;
    biomatch_doc: string;
    biomatch_selfie: string;
  };
  isRetryBiomatch?: boolean;
  guestId?: string;
};

function ConfirmIdentityResultScreen() {
  const {t} = useTranslation();
  const history = useHistory();
  const location = useLocation<LocationTypes>();
  const reservationDetails = useComputedReservationDetails();
  const {updateBiomatchSelfieGuest} = useBiomatchUpdateGuest();
  const faceMatchPercentage = getFaceMatchPercentage(
    location.state?.facesComparingResult?.distance,
  );
  const isMatch = location.state?.facesComparingResult?.is_match;

  const areBiomatchRetriesEnabled = Boolean(
    reservationDetails.reservation?.are_biomatch_retries_enabled,
  );

  const isPhotoVerificationDisabled = Boolean(
    reservationDetails.reservation.is_biomatch_photo_verification_disabled,
  );

  const goBackToDocument = React.useCallback(() => {
    history.push('document', location.state);
  }, [history, location.state]);

  const goBackToSelfie = React.useCallback(() => {
    history.push('identity', location.state);
  }, [history, location.state]);

  const getBiometricResultPayload = React.useCallback(() => {
    const state = location.state;
    return {
      biomatch_passed: isMatch || state?.biomatch?.biomatch_passed,
      biomatch_doc: state?.documentPhoto || state?.biomatch?.biomatch_doc,
      biomatch_selfie: state?.selfiePhoto || state?.biomatch?.biomatch_selfie,
    };
  }, [isMatch, location.state]);

  React.useEffect(() => {
    const hasPersistedState =
      location.state?.facesComparingResult &&
      ((location.state.documentPhoto && location.state.selfiePhoto) ||
        location.state.biomatch);

    if (!hasPersistedState) {
      goBackToDocument();
    }
  }, [goBackToDocument, location.state]);

  const getPersistedRetryState = () => {
    return location.state;
  };

  const retryMatching = () => {
    history.push('document', getPersistedRetryState());
  };

  const getPersistedState = () => {
    const biomatchPayload = getBiometricResultPayload();
    return {
      ...location.state,
      documentCheckId: undefined,
      documentPhoto: undefined,
      selfiePhoto: undefined,
      document: {},
      biomatch: {
        ...biomatchPayload,
      },
    };
  };

  const goNext = () => {
    const state = getPersistedState();

    if (state.isRetryBiomatch) {
      updateBiomatchSelfieGuest();
      return history.push('/biomatch-results', state);
    }
    if (reservationDetails.isDocScanDisabled) {
      history.push('/form/add', state);
    } else {
      history.push('/form/type', state);
    }
  };

  const getTitle = () => {
    if (isMatch) {
      return (
        <>
          {t('identity_verification_success_title')}
          <TitleSuccessIcon src={successSmallIcon} alt="Green check mark" />
        </>
      );
    }
    return (
      <>
        {t('pending')}
        <TitleErrorIcon src={errorSmallIcon} alt="Exclamation mark" />
      </>
    );
  };

  const getSubtitle = () => {
    if (!areBiomatchRetriesEnabled && !isMatch) {
      return t('identetify_verification_error_can_proceed');
    }

    if (isMatch) {
      return t('only_identity_verification_success_text');
    }

    return t('retry_identity_verification_error_text');
  };

  return (
    <>
      <Header
        {...getIdentityVerificationHeaderProps(reservationDetails)}
        hideBackButton={false}
        activeStep={2}
        onBack={goBackToSelfie}
        title={t('verify_your_identity')}
      />
      <Content>
        <Image src={scanningIcon} alt="Two persons" />
        <Title>{getTitle()}</Title>
        <Subtitle>{getSubtitle()}</Subtitle>
        <ScanResultContainer>
          <DocumentResultArea>
            <ResultImage
              mirror={!isMobileOrTablet()}
              image={location.state?.documentPhoto}
            />
            {breakIntoLines(t('passport_photo'))}
          </DocumentResultArea>
          <ProgressArea>
            <ProgressCircle
              progress={isMatch ? faceMatchPercentage : 0}
              errorText={
                !isMatch && !isPhotoVerificationDisabled ? t('error').toUpperCase() : ''
              }
              warningText={!isMatch && isPhotoVerificationDisabled ? t('pending') : ''}
              label={t('matching')}
            />
          </ProgressArea>
          <SelfieResultArea>
            <ResultImage mirror image={location.state?.selfiePhoto} />
            {breakIntoLines(t('selfie_photo'))}
          </SelfieResultArea>
        </ScanResultContainer>
        <ButtonWrapper>
          <Button data-testid="submit-btn" label={t('next')} onClick={goNext} />
        </ButtonWrapper>
        <TryButton label={t('try_again')} onClick={retryMatching} link />
      </Content>
    </>
  );
}

export {ConfirmIdentityResultScreen};
