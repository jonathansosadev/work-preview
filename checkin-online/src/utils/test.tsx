import React from 'react';
import {render} from '@testing-library/react';
import api from '../api';
import {createMemoryHistory} from 'history';
import useSWR from 'swr';
import {Router} from 'react-router-dom';
import {ReservationContext} from '../context/reservation';
import {ComputedReservationDetailsProvider} from '../context/computedReservationDetails';

const FAKE_RESERVATION = {
  data: {
    id: 1,
    housing: {
      id: 'testId',
      picture: {
        src: '',
      },
      location: {
        country: {
          code: 'ES',
        },
      },
    },
  },
  isError: false,
  isLoading: false,
  refreshReservation: () => {
    return api.reservations.getOne('');
  },
  patchReservation: () => {
    return api.reservations.patch('', {});
  },
};

function getNextReservation(reservation?: any) {
  if (!reservation) {
    return FAKE_RESERVATION;
  }

  return {
    ...FAKE_RESERVATION,
    ...reservation,
  };
}

function renderWithProviders(
  ui: React.ReactNode,
  reservation?: any,
  locationState?: any,
) {
  const nextReservation = getNextReservation(reservation);
  const history = createMemoryHistory();
  history.push('/', locationState || {});

  const utils = render(
    <Router history={history}>
      <ReservationContext.Provider value={nextReservation}>
        <ComputedReservationDetailsProvider>{ui}</ComputedReservationDetailsProvider>
      </ReservationContext.Provider>
    </Router>,
  );

  const rerenderWithProviders = (rerenderUi: React.ReactNode = ui, reservation?: any) => {
    const nextReservation = getNextReservation(reservation);
    return utils.rerender(
      <Router history={history}>
        <ReservationContext.Provider value={nextReservation}>
          <ComputedReservationDetailsProvider>
            {rerenderUi}
          </ComputedReservationDetailsProvider>
        </ReservationContext.Provider>
      </Router>,
    );
  };

  return {
    ...utils,
    history,
    rerenderWithProviders,
  };
}

function changeReservationCountryCode(reservation: any = {}, code = '') {
  return {
    ...reservation,
    data: {
      ...reservation.data,
      housing: {
        ...reservation.housing,
        location: {
          ...reservation.location,
          country: {
            ...reservation.country,
            code,
          },
        },
      },
    },
  };
}

function mockSWR(
  mocks: {
    url: string;
    response: {data: any; error?: any};
  }[] = [],
) {
  const swrBase = {
    data: null,
    error: null,
    isValidating: false,
    mutate: jest.fn(),
  };

  // @ts-ignore
  useSWR.mockImplementation((requestUrl: string | (() => string | null) = '') => {
    if (!requestUrl) {
      return {...swrBase};
    }

    const urlMock = mocks.find(mock => {
      if (typeof requestUrl === 'function') {
        const fnResult = requestUrl();
        if (!fnResult) {
          return {...swrBase};
        }

        return fnResult.includes(mock.url);
      } else {
        return requestUrl.includes(mock.url);
      }
    });

    if (urlMock) {
      return {
        ...swrBase,
        ...urlMock.response,
      };
    }

    return {...swrBase};
  });
}

export {FAKE_RESERVATION, renderWithProviders, changeReservationCountryCode, mockSWR};
