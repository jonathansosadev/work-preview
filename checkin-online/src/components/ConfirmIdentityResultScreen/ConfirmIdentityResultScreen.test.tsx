import React from 'react';
import {renderWithProviders} from '../../utils/test';
import {fireEvent} from '@testing-library/react';
import {ConfirmIdentityResultScreen} from './ConfirmIdentityResultScreen';

test('displays success or error based on match', async () => {
  const reservation = {
    data: {
      id: 'test id',
      guest_group: {
        leader_id: 'test',
      },
    },
  };

  const locationState = {
    facesComparingResult: {
      is_match: false,
    },
    isVerifying: true,
    documentPhoto: 'doc',
    selfiePhoto: 'selfie',
  };

  const {getByTestId, history, queryByAltText, findByText} = renderWithProviders(
    <ConfirmIdentityResultScreen />,
    reservation,
    locationState,
  );
  const historySpy = jest.spyOn(history, 'push');

  let successTitleIcon = queryByAltText(/green check mark/i);
  expect(successTitleIcon).not.toBeInTheDocument();
  let errorTitleIcon = queryByAltText(/exclamation mark/i);
  expect(errorTitleIcon).toBeInTheDocument();

  let submitButton = getByTestId(/submit-btn/i);
  expect(submitButton).toHaveTextContent(/try_again/i);
  fireEvent.click(submitButton);

  history.push('/', {
    ...locationState,
    facesComparingResult: {
      is_match: true,
    },
  });
  successTitleIcon = queryByAltText(/green check mark/i);
  expect(successTitleIcon).toBeInTheDocument();
  errorTitleIcon = queryByAltText(/exclamation mark/i);
  expect(errorTitleIcon).not.toBeInTheDocument();

  submitButton = await findByText(/next/i);
  fireEvent.click(submitButton);

  expect(historySpy.mock.calls[2]).toMatchSnapshot();
});
