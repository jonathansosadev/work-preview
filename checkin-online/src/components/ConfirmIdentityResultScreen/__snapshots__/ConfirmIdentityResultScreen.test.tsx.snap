// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays success or error based on match 1`] = `
Array [
  "/form/add",
  Object {
    "biomatch": Object {
      "biomatch_doc": "doc",
      "biomatch_passed": true,
      "biomatch_selfie": "selfie",
    },
    "document": Object {},
    "documentCheckId": undefined,
    "documentPhoto": undefined,
    "isVerifying": true,
    "facesComparingResult": Object {
      "is_match": true,
    },
    "selfiePhoto": undefined,
  },
]
`;
